src/library/scala/concurrent/impl/ExecutionContextImpl.scala

```diff
@@ -53,39 +52,42 @@ private[concurrent] object ExecutionContextImpl {
       thread
     }
 
-    // As per ThreadFactory contract newThread should return `null` if cannot create new thread.
-    def newThread(runnable: Runnable): Thread =
-      if (reserveThread())
-        wire(new Thread(new Runnable {
-          // We have to decrement the current thread count when the thread exits
-          override def run() = try runnable.run() finally deregisterThread()
-        })) else null
+    def newThread(runnable: Runnable): Thread = wire(new Thread(runnable))
```

Comment:
Can this (extending `TheadFactory`) be removed? Somewhat orthogonal to this PR, but once you're rewriting the implementation, might as well remove it if it makes sense.

Reply:
I can remove it in a separate commit in the same PR—it is there for historical reasons.

Reply:
done